subsys/net/lib/zperf/zperf_tcp_receiver.c

```diff
@@ -99,6 +96,139 @@ static void tcp_received(const struct sockaddr *addr, size_t datalen)
 	}
 }
 
+static void tcp_session_error_report(void)
+{
+	if (tcp_session_cb != NULL) {
+		tcp_session_cb(ZPERF_SESSION_ERROR, NULL, tcp_user_data);
+	}
+}
+
+static int tcp_recv_data(struct net_socket_service_event *pev)
+{
+	static uint8_t buf[TCP_RECEIVER_BUF_SIZE];
+	int i, ret = 0;
+	int family, sock;
+	struct sockaddr addr_incoming_conn;
+	socklen_t optlen = sizeof(int);
+	socklen_t addrlen = sizeof(struct sockaddr);
+
+	if (tcp_server_stop) {
+		ret = -ENOENT;
+		goto cleanup;
+	}
+
+	if ((pev->event.revents & ZSOCK_POLLERR) ||
+	    (pev->event.revents & ZSOCK_POLLNVAL)) {
+		(void)zsock_getsockopt(pev->event.fd, SOL_SOCKET,
+				       SO_DOMAIN, &family, &optlen);
+		NET_ERR("TCP receiver IPv%d socket error",
+			family == AF_INET ? 4 : 6);
+		goto error;
+	}
+
+	if (!(pev->event.revents & ZSOCK_POLLIN)) {
+		return 0;
+	}
+
+	/* What is the index to first accepted socket */
+	i = SOCK_ID_IPV6_LISTEN + 1;
+
+	/* Check first if we need to accept a connection */
+	if (fds[SOCK_ID_IPV4_LISTEN].fd == pev->event.fd ||
+	    fds[SOCK_ID_IPV6_LISTEN].fd == pev->event.fd) {
+		sock = zsock_accept(pev->event.fd,
+				    &addr_incoming_conn,
+				    &addrlen);
+		if (sock < 0) {
+			(void)zsock_getsockopt(pev->event.fd, SOL_SOCKET,
+					       SO_DOMAIN, &family, &optlen);
+			NET_ERR("TCP receiver IPv%d accept error",
+				family == AF_INET ? 4 : 6);
+			goto error;
+		}
+
+		for (; i < SOCK_ID_MAX; i++) {
+			if (fds[i].fd < 0) {
+				break;
+			}
+		}
+
+		if (i == SOCK_ID_MAX) {
+			/* Too many connections. */
+			NET_ERR("Dropping TCP connection, reached maximum limit.");
+			zsock_close(sock);
+		} else {
+			fds[i].fd = sock;
+			fds[i].events = ZSOCK_POLLIN;
+			memcpy(&sock_addr[i], &addr_incoming_conn, addrlen);
+
+			(void)net_socket_service_register(&svc_tcp, fds,
+							  ARRAY_SIZE(fds),
+							  NULL);
+		}
+
+	} else {
+		ret = zsock_recv(pev->event.fd, buf, sizeof(buf), 0);
+		if (ret < 0) {
+			(void)zsock_getsockopt(pev->event.fd, SOL_SOCKET,
+					       SO_DOMAIN, &family, &optlen);
+			NET_ERR("recv failed on IPv%d socket (%d)",
+				family == AF_INET ? 4 : 6,
+				errno);
+			tcp_session_error_report();
+			/* This will close the zperf session */
+			ret = 0;
+		}
+
+		for (; i < SOCK_ID_MAX; i++) {
+			if (fds[i].fd == pev->event.fd) {
```

Comment:
Would it be possible that we won't find FD on the list? Just wonder if we should check if `i == SOCK_ID_MAX` after the loop, and possibly handle it as an error?

Reply:
Good point, there was a check in the listen socket branch but the check was missing here, I will add one.